### DISCUSSION TITLE
Filter on multiple possible senders

### DISCUSSION THREAD

**Discussion by IT-VBFK** [jstedfast/MailKit#1769] [Q&A] [answered]
Hi

I have a usecase, where I want to automatically fetch IMAP mails from an account, but because it is auto-fetching I only want to download and process mails from permitted senders which are unread.
So I tried this

```csharp
string[] permittedSenders = // passed in via .env
var query = new SearchQuery();

foreach(var s in permittedSenders)
{
    query = query.Or(SearchQuery.FromContains(s))
}
client.Inbox.Search(query.And(SearchQuery.NotSeen))

//...
```
But apparently unseen mails only from the last permitted sender are returned. 

Am I missing something?

**Comment by jstedfast** [comment, marked as answer]
I would do something more like this:

```csharp
string[] permittedSenders = // passed in via .env
var query = SearchQuery.FromContains(permittedSenders[0]);

for (int i = 1; i < permittedSenders.Count; i++)
{
    query = query.Or(SearchQuery.FromContains(permittedSenders[i]))
}
client.Inbox.Search(query.And(SearchQuery.NotSeen))
```

Keep in mind that `new SearchQuery()` is saying "ALL messages" and so the query you are using right now is "(ALL messages -OR- messages from permittedSenders[0] -OR- permittedSenders[1] ...) -AND- NotSeen"

I'm not sure why you are getting the results you are getting because to me, it would make more sense that you are getting all unseen messages.

That said, the query in my example should be correct. Hopefully you get the results you are expecting.

If not, I would recommend getting a protocol log (see the FAQ.md in the root of the MailKit repo) and then paste the SEARCH command here so we can see what is going on (obviously feel free to redact the email addresses).

**Reply by IT-VBFK** [reply]
> Keep in mind that new SearchQuery() is saying "ALL messages" and so the query you are using right now is "(ALL messages -OR- messages from permittedSenders[0] -OR- permittedSenders[1] ...) -AND- NotSeen"

Ahh.. that is good to know, thanks

I have to admit, that in the time of writing this post I were away from my dev machine, so I couldn't see that windows has rebooted (and also the docker container where this process was running), so in fact: it returned obviously nothing 🙈

My first guess was (because I expected the container up and running), that this must have something to do with the query, because the mails were coming from first permitted sender in the array... so... sorry for the `MailKit`- blame 🙏

**Reply by jstedfast** [reply]
No worries. Hopefully that works out for you.
